internal/cmd/io_binary.go

```diff
@@ -0,0 +1,234 @@
+package cmd
+
+import (
+	"context"
+	"fmt"
+	"io"
+	"net"
+	"net/url"
+	"os/exec"
+	"sync"
+	"time"
+
+	"github.com/Microsoft/go-winio"
+	"github.com/Microsoft/hcsshim/internal/log"
+	"github.com/containerd/containerd/namespaces"
+	"github.com/pkg/errors"
+)
+
+const (
+	binaryPipeFmt        = `\\.\pipe\binary-%s-%s`
+	binaryCmdWaitTimeout = 5 * time.Second
+)
+
+func newBinaryCmd(ctx context.Context, uri *url.URL, id string, ns string) *exec.Cmd {
+	var args []string
+	for k, vs := range uri.Query() {
+		args = append(args, k)
+		if len(vs) > 0 && vs[0] != "" {
+			args = append(args, vs[0])
+		}
+	}
+
+	execPath := uri.Path
+	if execPath == "" {
+		execPath = uri.Host
+	}
+
+	cmd := exec.CommandContext(ctx, execPath, args...)
+	cmd.Env = append(cmd.Env,
+		"CONTAINER_ID="+id,
+		"CONTAINER_NAMESPACE="+ns,
+	)
+
+	return cmd
+}
+
+type binaryIO struct {
+	cmd *exec.Cmd
+
+	binaryCloser sync.Once
+
+	stdin, stdout, stderr string
+
+	sout, serr io.ReadWriteCloser
+	soutCloser sync.Once
+}
+
+// NewBinaryIO starts a binary logger
+func NewBinaryIO(ctx context.Context, id string, uri *url.URL) (_ UpstreamIO, err error) {
+	ns, err := namespaces.NamespaceRequired(ctx)
+	if err != nil {
+		return nil, err
+	}
+
+	var sout, serr, w io.ReadWriteCloser
+
+	stdoutPipe := fmt.Sprintf(binaryPipeFmt, id, "stdout")
+	sout, err = openNPipe(stdoutPipe)
+	if err != nil {
+		return nil, err
+	}
+
+	stderrPipe := fmt.Sprintf(binaryPipeFmt, id, "stderr")
+	serr, err = openNPipe(stderrPipe)
+	if err != nil {
+		return nil, err
+	}
+
+	waitPipe := fmt.Sprintf(binaryPipeFmt, id, "wait")
+	w, err = openNPipe(waitPipe)
+	if err != nil {
+		return nil, err
+	}
+
+	defer w.Close()
+
+	cmd := newBinaryCmd(ctx, uri, id, ns)
+	cmd.Env = append(cmd.Env,
```

Comment:
Refer to this: https://github.com/microsoft/hcsshim/pull/896/files#r532625214

Don't feel too strongly about it but it seems easier to read/reason with to me to just pass in the entirety of the env block in instead of constructing half of it inside `newBinaryCmd` and then adding more right after.

Reply:
makes sense. I was following the upstream example, but in their case `newBinaryCmd` is public, so it makes sense there to have some default env when people want to write extensions or what not, but in our case it doesn't really make sense.